### PR TITLE
WIP, VIZ: plot_compare_evokeds can plot epochs

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
As discussed in #5812: plot_compare_evokeds can take dicts of epochs. This allows:

- for a dict of epochs: plotting the confidence interval within a subject/across trials
- for a dict of lists of epochs: automatically selecting conditions from the epochs without having to construct the evoked objects

The API is unchanged. Only now code like this is legal:

```plot_compare_evokeds({"aud": epochs, "vis": epochs})```

@palday @mmagnuski

### CODE REVIEW

**Comment by jona-sassenhagen** [comment]
... the next step would be adding this as a method to Epochs. I.e., 

```epochs.plot_compare_evokeds()```

would plot one line per event_id.

```epochs.plot_compare_evokeds(("aud", "viz"))```

would be equivalent to `plot_compare_evokeds({"viz": epochs, "aud": epochs})`.

Would be nice for fast evaluation of simple contrasts, and, again, the API comes for free as it'd just be adding the existing function as a method to Epochs.

File: mne/viz/evoked.py [hunk 1]
```diff
@@ -1549,9 +1549,26 @@ def _check_loc_legal(loc, what='your choice', default=1):
     return loc_
 
 
+def _get_evokeds_from_epochs(evokeds, Epochs):
+    """Transform a dict of lists of Epochs into a dict of lists of Evokeds."""
+    if all(isinstance(content, Epochs) for content in evokeds.values()):
+        evokeds = {key: list(epochs[0][key].iter_evoked())
```
Comment (jona-sassenhagen) [inline mne/viz/evoked.py]: note to self: I think this 0 is wrong, check tmrw

File: mne/viz/evoked.py [hunk 2]
```diff
@@ -1565,15 +1582,34 @@ def _format_evokeds_colors(evokeds, cmap, colors):
...
+    for condition in evokeds.values():
+        if isinstance(condition, (Epochs, Evoked)):
+            all_types.append(type(condition))
+        else:
+            for item in condition:
+                all_types.append(type(item))
+    all_types = set(all_types)
+    if len(all_types) != 1:
+        raise ValueError("All instances in `evokeds` must be of the same type "
+                         "(Evokeds OR Epochs), got " + str(all_types))
```
Comment (agramfort) [inline mne/viz/evoked.py]: please use/write a check function in check.py

also I am not super happy to see something called evokeds that can contains epochs. It should ne renamed to "inst"

**Comment by jona-sassenhagen** [comment]
Done, thanks @agramfort .

Still WIP, needs an example etc.

**Comment by jona-sassenhagen** [comment]
I think this is elegant:

```
from mne.datasets import testing
data_path = testing.data_path()
fname = data_path + "/EEGLAB/test_raw.set"
montage = data_path + "/EEGLAB/test_chans.locs"

event_id = {"rt":1, "square":2}
eog = {"FPz", "EOG1", "EOG2"}
raw = mne.io.eeglab.read_raw_eeglab(fname, preload=True, eog=eog, montage=montage)
events, event_id = mne.events_from_annotations(raw)
picks = mne.pick_types(raw.info, eeg=True)
epochs = mne.Epochs(raw, events,  event_id, tmax=.7, picks=picks)

epochs.plot_compare_evokeds(picks=4);
```

unknown-223

You can also provide a `conditions` arg, with a list of the events you want to plot.
(Alternatively, this could be skipped, as it approximates `epochs[conds].plot_compare_evokeds()`)

**Comment by jona-sassenhagen** [comment]
Or this: 

```
import os.path as op
import mne

data_path = op.join(mne.datasets.sample.data_path(), 'MEG', 'sample')
raw = mne.io.read_raw_fif(
    op.join(data_path, 'sample_audvis_raw.fif'), preload=True)
raw.filter(None, 30, fir_design='firwin')
event_id = {'auditory/left': 1, 'auditory/right': 2, 'visual/left': 3, 'visual/right': 4}
events = mne.read_events(op.join(data_path, 'sample_audvis_raw-eve.fif'))
epochs = mne.Epochs(raw, events, event_id=event_id, tmin=-0.2, tmax=.5)

epochs.plot_compare_evokeds(picks=4, ci=.68, split_legend=True,
                            colors={"left": "C0", "right": "C1"},
                            linestyles={"auditory": "-", "visual": ":"});
```

unknown-225

**Comment by jona-sassenhagen** [comment]
@agramfort What do you think?

@mmagnuski useful?

**Comment by jona-sassenhagen** [comment]
(The CI errors are due to the lack of docstring for the method.)

**Comment by mmagnuski** [comment]
@jona-sassenhagen Sure, looks very useful. In general I perfer separation of computation (here - erp) and plotting but in this case the convenience wins for me (and the computation is trivial and super-fast anyway so dropping intermediate results does not hurt anyone).

**Comment by jona-sassenhagen** [comment]
@agramfort are you in principle ok with this?

**Comment by palday** [comment]
Would it make sense to have the `Epochs` method be a bit smarter for hierarchical `event_id`s? Right now, they are treated as flat and assigned a different color, but for 1-deep nesting it would be straightforward to do color for the outer level and linetype for the inner level.

**Comment by jona-sassenhagen** [comment]
I thought about that, but my hunch is this would be too much magic for @agramfort :)

Note we enforce zero structure on event ids. Although maybe this should change - we could more openly suggest, e.g., an ordering scheme, and implicitly make use of it.

**Comment by palday** [comment]
Yeah, I'm also somewhat torn about the necessity of having the `Epochs` method, even though / because it's just a wrapper for the `viz` method.

**Comment by jona-sassenhagen** [comment]
@agramfort ..?

It seems @nbara and @mmagnuski like it ...

Maybe @sappelhoff has an opinion too.

**Comment by agramfort** [comment]
no strong feeling here but I would suggest to discuss during the sprint next month.

**Comment by sappelhoff** [comment]
+1 looks very useful for me during initial exploration in a notebook :-)

**Comment by jona-sassenhagen** [comment]
This probably became irrecoverably conflicted with Dan's big refactor. I am considering closing and writing new from scratch.

**Pull request closed without merging**